helix-lsp/src/lib.rs

```diff
@@ -202,34 +202,120 @@ pub mod util {
         doc: &Rope,
         selection: &Selection,
         edit: lsp::TextEdit,
+        is_snippet: bool,
         offset_encoding: OffsetEncoding,
-    ) -> Transaction {
-        let replacement: Option<Tendril> = if edit.new_text.is_empty() {
-            None
-        } else {
-            Some(edit.new_text.into())
-        };
+    ) -> (Transaction, Option<Vec<(u32, Selection)>>) {
+        // We need to process the snippet input to:
+        //  1. Remove and store the snippet tab stops
+        //  2. Offset each newline with N whitespace characters, so that the whole snippet is
+        //     correctly indented
+        //
+        // Tab stops: A snippet can define tab stops and placeholders with $1, $2 and ${3:foo}.
+        // $0 defines the final tab stop, it defaults to the end of the snippet.
+        fn process_snippet(
+            snippet: &str,
+            newline_offset: usize,
+            selection_range: &Range,
+            start_offset: i128,
+        ) -> (String, Vec<(u32, Selection)>) {
+            let mut new_snippet = String::with_capacity(snippet.len());
+            let mut tab_stops = Vec::new();
+
+            let whitespace_offset = str::repeat(" ", newline_offset);
+            for (i, mut l) in snippet.lines().enumerate() {
+                if i != 0 {
+                    new_snippet.push_str("\n");
```

Comment:
You might want to use the file line ending form `Editor` instead of the hardcoded `"\n"`. This would make it easier for cross-platform compatibility.